File: packages/adapters/hermes/selfmem_canary_standalone_smoke.py
#!/usr/bin/env python3
"""Standalone smoke test for the selfmem Hermes canary provider.

This fakes the minimal Hermes MemoryProvider import so the canary can be tested
outside a live Hermes checkout. The live VM should still run
`selfmem_canary_smoke.py` from the real Hermes repo before promotion.
"""

from __future__ import annotations

import importlib.util
import json
import re
import sys
import tempfile
import types
from pathlib import Path


ROOT = Path(__file__).resolve().parents[3]
PROVIDER_PATH = ROOT / "packages" / "adapters" / "hermes" / "selfmem_canary" / "__init__.py"


def main() -> None:
    fake_agent = types.ModuleType("agent")
    fake_memory_provider = types.ModuleType("agent.memory_provider")

    class MemoryProvider:
        pass

    fake_memory_provider.MemoryProvider = MemoryProvider
    sys.modules["agent"] = fake_agent
    sys.modules["agent.memory_provider"] = fake_memory_provider

    spec = importlib.util.spec_from_file_location("selfmem_canary_standalone", PROVIDER_PATH)
    if spec is None or spec.loader is None:
        raise SystemExit(f"Could not load provider: {PROVIDER_PATH}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    with tempfile.TemporaryDirectory(prefix="selfmem-hermes-smoke-") as tmp:
        hermes_home = Path(tmp) / ".hermes"
        hermes_home.mkdir(parents=True, exist_ok=True)
        provider = module.SelfmemCanaryProvider()
        provider.initialize(
            "standalone-smoke",
            hermes_home=str(hermes_home),
            agent_identity="standalone-agent",
            supermemory_container="hermes_standalone_source",
        )
        provider._supermemory_read_through = True
        provider._supermemory_key = "test-read-through-key"
        provider._provider_mode = f"{provider._provider_mode}+supermemory-read-through-test"

        def fake_supermemory_search(query: str, limit: int):
            return [{
                "id": "supermemory:remote-smoke",
                "content": "Remote Supermemory history remains searchable while new memories write locally.",
                "score": 0.91,
                "provider_mode": "supermemory_read_through",
                "memory_source": "supermemory_read_through",
                "metadata": {
                    "source_supermemory_container": "hermes_standalone_source",
                },
            }]

        provider._search_supermemory = fake_supermemory_search
        tool_names = [schema["name"] for schema in provider.get_tool_schemas()]
        alias_store = json.loads(provider.handle_tool_call("supermemory_store", {
            "content": "LCM lifecycle memory should be distilled into local RecallWeave before any future Supermemory sync.",
        }))
        alias_search = json.loads(provider.handle_tool_call("supermemory_search", {
            "query": "LCM lifecycle local RecallWeave sync remote Supermemory history",
            "limit": 5,
        }))
        provider.sync_turn(
            "remember the Hermes canary must preserve LCM lifecycle events locally",
            "I will store the durable local memory lifecycle rule.",
            session_id="standalone-smoke",
        )
        provider.sync_turn(
            [{"type": "text", "text": "remember list-shaped Hermes lifecycle payloads must not crash selfmem"}],
            [{"type": "text", "text": "I will handle structured content safely."}],
            session_id="standalone-smoke",
        )
        prefetch = provider.prefetch("What is the LCM lifecycle memory rule?", session_id="standalone-smoke")
        skipped_prefetch = provider.prefetch("heartbeat diagnostic status check ping", session_id="standalone-smoke")
        status_like_prefetch = provider.prefetch("What is the status decision for Hermes memory retrieval?", session_id="standalone-smoke")
        pre_compress = provider.on_pre_compress([
            {"role": "user", "content": [{"type": "text", "text": "LCM pre-compress should be tracked by selfmem."}]},
        ])
        provider.on_memory_write("remember", "MEMORY.md", "Built-in Hermes memory writes should mirror into local RecallWeave.")
        provider.on_session_end([
            {"role": "user", "content": "session ending after LCM lifecycle smoke"},
            {"role": "assistant", "content": "done"},
        ])
        status = json.loads(provider.handle_tool_call("supermemory_status", {}))
        provider.shutdown()

        memories = Path(status["memories_path"]).read_text(encoding="utf-8", errors="ignore")
        container_map = json.loads(Path(status["container_map_path"]).read_text(encoding="utf-8", errors="ignore"))
        trace = Path(status["trace_path"]).read_text(encoding="utf-8", errors="ignore")
        lossless = Path(status["lossless_path"]).read_text(encoding="utf-8", errors="ignore")
        raw = Path(status["raw_path"]).read_text(encoding="utf-8", errors="ignore")
        trace_events = [json.loads(line) for line in trace.splitlines() if line.strip()]
        search_trace = next(
            (
                event
                for event in trace_events
                if event.get("event") == "search" and event.get("data", {}).get("supermemory_attempted") is True
            ),
            {},
        )
        search_trace_data = search_trace.get("data", {})
        store_traces = [event for event in trace_events if event.get("event") == "store"]
        leaks = count_leaks("\n".join([memories, trace, lossless, raw, prefetch]))
        output = {
            "ok": True,
            "toolAliasCoverage": all(name in tool_names for name in [
                "supermemory_store",
                "supermemory_search",
                "supermemory_forget",
                "supermemory_profile",
                "supermemory_status",
            ]),
            "aliasStoreSuccess": bool(alias_store.get("success")),
            "aliasSearchResultCount": len(alias_search.get("results", [])),
            "hybridSearchCovered": any(item.get("memory_source") == "supermemory_read_through" for item in alias_search.get("results", [])),
            "sourceSupermemoryContainer": status.get("source_supermemory_container"),
            "localContainer": status.get("local_container"),
            "adapterContractCovered": status.get("adapter_contract", {}).get("name") == "recallweave-selfmem-canary"
            and status.get("adapter_contract", {}).get("strictCanaryContract") == "v1"
            and status.get("adapter_contract", {}).get("searchLatencyInstrumentation") is True
            and status.get("adapter_contract", {}).get("storeLatencyInstrumentation") is True
            and container_map.get("adapter_contract", {}).get("strictCanaryContract") == "v1",
            "boundedReadThroughPolicyCovered": status.get("search_policy") == "local_first_then_bounded_supermemory_read_through"
            and status.get("recall_policy", {}).get("remote_read_through") == "explicit_history_intent_or_thin_local_results",
            "searchLatencyInstrumentationCovered": float(search_trace_data.get("elapsed_ms") or 0) > 0
            and float(search_trace_data.get("local_elapsed_ms") or 0) > 0
            and float(search_trace_data.get("remote_elapsed_ms") or 0) > 0,
            "storeLatencyInstrumentationCovered": bool(store_traces)
            and all(float(event.get("data", {}).get("elapsed_ms") or 0) > 0 for event in store_traces),
            "storeLatencySampleCount": sum(
                1 for event in store_traces if float(event.get("data", {}).get("elapsed_ms") or 0) > 0
            ),
            "prefetchHasContext": bool(prefetch.strip()),
            "maintenanceRecallGateCovered": skipped_prefetch == "" and "prefetch_skipped" in trace,
            "statusLikeRecallCovered": bool(status_like_prefetch.strip()),
            "preCompressReturned": bool(pre_compress.strip()),
            "lcmLifecycleCovered": all(token in lossless for token in [
                "sync_turn_candidate",
                "pre_compress",
                "memory_write_candidate",
                "session_end",
            ]),
            "rawAuditCovered": "sync_turn_raw" in raw and "session_end_raw" in raw,
            "listPayloadCovered": "list-shaped Hermes lifecycle payloads" in memories,
            "distilledNotRawDefault": "role: user" not in memories,
            "privacyLeakCount": leaks,
            "memoryLineCount": len([line for line in memories.splitlines() if line.strip()]),
            "traceLineCount": len([line for line in trace.splitlines() if line.strip()]),
            "losslessLineCount": len([line for line in lossless.splitlines() if line.strip()]),
            "rawLineCount": len([line for line in raw.splitlines() if line.strip()]),
        }
        print(json.dumps(output, indent=2, sort_keys=True))
        if not all([
            output["toolAliasCoverage"],
            output["adapterContractCovered"],
            output["aliasStoreSuccess"],
            output["aliasSearchResultCount"] > 0,
            output["hybridSearchCovered"],
            output["boundedReadThroughPolicyCovered"],
            output["searchLatencyInstrumentationCovered"],
            output["storeLatencyInstrumentationCovered"],
            output["sourceSupermemoryContainer"] == "hermes_standalone_source",
            output["localContainer"] == "selfmem_hermes_standalone_source",
            output["prefetchHasContext"],
            output["maintenanceRecallGateCovered"],
            output["statusLikeRecallCovered"],
            output["preCompressReturned"],
            output["lcmLifecycleCovered"],
            output["rawAuditCovered"],
            output["listPayloadCovered"],
            output["distilledNotRawDefault"],
            output["privacyLeakCount"] == 0,
        ]):
            raise SystemExit(1)


def count_leaks(text: str) -> int:
    patterns = [
        r"</?private>",
        r"pa-[A-Za-z0-9_-]{20,}",
        r"nvapi-[A-Za-z0-9_-]{20,}",
        r"jina_[A-Za-z0-9_-]{20,}",
        r"sk-ant-[A-Za-z0-9_-]{20,}",
        r"sm_[A-Za-z0-9_-]{20,}",
        r"AIza[0-9A-Za-z_-]{20,}",
    ]
    return sum(1 for pattern in patterns if re.search(pattern, text))


if __name__ == "__main__":
    main()

File: packages/adapters/openclaw/selfmem_canary_standalone_smoke.mjs
#!/usr/bin/env node
import { execFileSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import pluginDefinition, { createSelfmemOpenClawCanary } from "./selfmem_canary/index.mjs";

const home = mkdtempSync(join(tmpdir(), "selfmem-openclaw-smoke-"));
const providerCalls = { embeddings: 0, rerank: 0, supermemory: 0 };
globalThis.fetch = async (url, options = {}) => {
  const href = String(url);
  const body = JSON.parse(String(options.body || "{}"));
  if (href.includes("/embeddings")) {
    providerCalls.embeddings += 1;
    const input = Array.isArray(body.input) ? body.input : [];
    return new Response(JSON.stringify({
      data: input.map((text, index) => ({
        index,
        embedding: embeddingFor(String(text)),
      })),
      usage: { total_tokens: input.join(" ").split(/\s+/).filter(Boolean).length },
    }), { status: 200, headers: { "content-type": "application/json" } });
  }
  if (href.includes("/rerank")) {
    providerCalls.rerank += 1;
    const documents = Array.isArray(body.documents) ? body.documents : [];
    return new Response(JSON.stringify({
      data: documents.map((document, index) => ({
        index,
        relevance_score: String(document).toLowerCase().includes("openclaw") ? 0.98 : 0.4,
      })).sort((a, b) => b.relevance_score - a.relevance_score),
      usage: { total_tokens: documents.join(" ").split(/\s+/).filter(Boolean).length },
    }), { status: 200, headers: { "content-type": "application/json" } });
  }
  providerCalls.supermemory += 1;
  return new Response(JSON.stringify({
    results: [{
      id: "remote-smoke",
      memory: "Remote Supermemory history remains searchable while OpenClaw writes new memories locally.",
      similarity: 0.91,
    }],
  }), { status: 200, headers: { "content-type": "application/json" } });
};
const plugin = createSelfmemOpenClawCanary({
  home,
  agentIdentity: "openclaw-standalone-agent",
  supermemoryContainer: "openclaw_standalone_source",
  supermemoryKey: "test-read-through-key",
  voyageKeys: ["test-voyage-key"],
});

const session = plugin.session_start({ id: "openclaw-smoke" });
const store = plugin.tools.supermemory_store({
  content: "OpenClaw selfmem canary should capture agent_end and retrieve before prompt build.",
});
const search = await plugin.tools.supermemory_search({
  query: "OpenClaw canary retrieve before prompt",
  limit: 5,
});
const pre = await plugin.before_prompt_build({ query: "What should OpenClaw selfmem canary do?" });
const skippedPre = await plugin.before_prompt_build({ query: "heartbeat diagnostic status check ping" });
const statusLikeRealPre = await plugin.before_prompt_build({ query: "What is the status decision for OpenClaw memory retrieval?" });
plugin.agent_end({ summary: "Decision: OpenClaw agent_end writes distilled local RecallWeave memory for this agent only." });
plugin.compression_checkpoint({
  event: "lcm_pre_compress",
  summary: "LCM compression checkpoint should preserve durable local memory before compaction.",
});
const status = plugin.tools.supermemory_status();

const envHome = mkdtempSync(join(tmpdir(), "selfmem-openclaw-state-dir-"));
writeFileSync(join(envHome, "profile-identity.json"), JSON.stringify({
  agent_identity: "state-dir-agent",
  source_supermemory_container: "state_dir_source",
  local_container: "selfmem_state_dir_source",
}, null, 2));
const oldStateDir = process.env.OPENCLAW_STATE_DIR;
process.env.OPENCLAW_STATE_DIR = envHome;
const stateDirPlugin = createSelfmemOpenClawCanary({
  supermemoryKey: "test-read-through-key",
  voyageKeys: ["test-voyage-key"],
});
const stateDirStatus = stateDirPlugin.tools.selfmem_status();
if (oldStateDir === undefined) delete process.env.OPENCLAW_STATE_DIR;
else process.env.OPENCLAW_STATE_DIR = oldStateDir;

const readOnlyHome = mkdtempSync(join(tmpdir(), "selfmem-openclaw-readonly-"));
const readOnlyPlugin = createSelfmemOpenClawCanary({ home: readOnlyHome });
const readOnlyStore = readOnlyPlugin.tools.selfmem_store({
  content: "This should not write because identity is unresolved.",
});
const readOnlyStatus = readOnlyPlugin.tools.selfmem_status();

const registered = { tools: [], events: [], services: [], memoryCapability: false };
pluginDefinition.register({
  pluginConfig: {
    home,
    agentIdentity: "openclaw-standalone-agent",
    sourceSupermemoryContainer: "openclaw_standalone_source",
    supermemoryReadKey: "test-read-through-key",
    voyageKeys: ["test-voyage-key"],
  },
  logger: { info() {}, warn() {}, error() {}, debug() {} },
  registerTool(tool, options) { registered.tools.push(options?.name || tool.name); },
  on(event, handler) { registered.events.push(event); registered[`handler_${event}`] = handler; },
  registerService(service) { registered.services.push(service.id); },
  registerMemoryCapability() { registered.memoryCapability = true; },
});

const traceText = readFileSync(join(home, "selfmem", "containers", "selfmem_openclaw_standalone_source", "trace.jsonl"), "utf8");
const rawText = readFileSync(join(home, "selfmem", "containers", "selfmem_openclaw_standalone_source", "raw_events.jsonl"), "utf8");
const traceEvents = traceText.split(/\n+/).filter(Boolean).map((line) => JSON.parse(line));
const searchTrace = traceEvents.find((item) => item.event === "search" && item.data?.supermemory_attempted === true);
const storeTraces = traceEvents.filter((item) => item.event === "store");
const audit = JSON.parse(execFileSync("python3", [
  new URL("./selfmem_audit.py", import.meta.url).pathname,
  "--home",
  home,
  "--json",
], { encoding: "utf8" }));

const output = {
  ok: true,
  sessionLocalContainer: session.local_container,
  sourceSupermemoryContainer: status.source_supermemory_container,
  localContainer: status.local_container,
  boundedReadThroughPolicyCovered: status.search_policy === "local_first_then_bounded_supermemory_read_through"
    && status.recall_policy?.remote_read_through === "explicit_history_intent_or_thin_local_results",
  searchLatencyInstrumentationCovered: Number(searchTrace?.data?.elapsed_ms || 0) > 0
    && Number(searchTrace?.data?.local_elapsed_ms || 0) > 0
    && Number(searchTrace?.data?.remote_elapsed_ms || 0) > 0,
  storeLatencyInstrumentationCovered: storeTraces.length > 0
    && storeTraces.every((item) => Number(item.data?.elapsed_ms || 0) > 0),
  storeLatencySampleCount: storeTraces.filter((item) => Number(item.data?.elapsed_ms || 0) > 0).length,
  aliasStoreSuccess: Boolean(store.success),
  aliasSearchResultCount: search.results.length,
  hybridSearchCovered: search.results.some((item) => item.memory_source === "supermemory_read_through"),
  voyageEnabled: Boolean(status.voyage_enabled),
  voyageCallsCovered: Number(status.usage?.embedding_calls || 0) > 0 && Number(status.usage?.rerank_calls || 0) > 0,
  embeddingCacheCovered: providerCalls.embeddings <= 5 && Number(status.usage?.embedding_cache_hits || 0) > 0,
  maintenanceRecallGateCovered: skippedPre === "" && traceText.includes("before_prompt_build_skipped"),
  statusLikeRecallCovered: statusLikeRealPre.includes("<selfmem-context>"),
  openclawStateDirCovered: stateDirStatus.local_container === "selfmem_state_dir_source" && existsSync(join(envHome, "selfmem", "containers", "selfmem_state_dir_source", "container-map.json")),
  identityPinCovered: stateDirStatus.agent_identity === "state-dir-agent" && stateDirStatus.identity_resolved === true && stateDirStatus.read_only === false,
  readOnlyCovered: readOnlyStatus.read_only === true && readOnlyStore.success === false,
  pluginEntryCovered: pluginDefinition.id === "selfmem_canary" && registered.tools.includes("supermemory_search") && registered.events.includes("before_prompt_build") && registered.memoryCapability,
  compressionCheckpointCovered: traceText.includes("compression_checkpoint") && rawText.includes("compression_checkpoint_raw"),
  auditCovered: audit.summary?.containerCount >= 1 && audit.summary?.privacyLeakCount === 0,
  beforePromptHasContext: pre.includes("<selfmem-context>"),
  lifecycleCovered: traceText.includes("session_start") && traceText.includes("before_prompt_build") && traceText.includes("agent_end"),
  rawAuditCovered: rawText.includes("agent_end_raw"),
  privacyLeakCount: countLeaks(`${traceText}\n${rawText}\n${pre}`),
};

console.log(JSON.stringify(output, null, 2));

if (!output.boundedReadThroughPolicyCovered || !output.searchLatencyInstrumentationCovered || !output.storeLatencyInstrumentationCovered || !output.aliasStoreSuccess || output.aliasSearchResultCount < 1 || !output.hybridSearchCovered || !output.voyageEnabled || !output.voyageCallsCovered || !output.embeddingCacheCovered || !output.maintenanceRecallGateCovered || !output.statusLikeRecallCovered || !output.openclawStateDirCovered || !output.identityPinCovered || !output.readOnlyCovered || !output.pluginEntryCovered || !output.compressionCheckpointCovered || !output.auditCovered || !output.beforePromptHasContext || !output.lifecycleCovered || !output.rawAuditCovered || output.privacyLeakCount !== 0) {
  process.exit(1);
}

function embeddingFor(text) {
  const lower = text.toLowerCase();
  return [
    lower.includes("openclaw") ? 1 : 0.1,
    lower.includes("memory") ? 1 : 0.1,
    lower.includes("remote") ? 1 : 0.1,
  ];
}

function countLeaks(text) {
  return [
    /<\/?private>/,
    /pa-[A-Za-z0-9_-]{20,}/,
    /nvapi-[A-Za-z0-9_-]{20,}/,
    /jina_[A-Za-z0-9_-]{20,}/,
    /sk-[A-Za-z0-9_-]{20,}/,
    /sm_[A-Za-z0-9_-]{20,}/,
    /AIza[0-9A-Za-z_-]{20,}/,
  ].filter((pattern) => pattern.test(text)).length;
}
